src/cargo/util/config.rs

```diff
@@ -153,10 +153,54 @@ impl Config {
 
     /// Get the path to the `cargo` executable
     pub fn cargo_exe(&self) -> CargoResult<&Path> {
-        self.cargo_exe.get_or_try_init(||
-            env::current_exe().and_then(|path| path.canonicalize())
-            .chain_err(|| "couldn't get the path to cargo executable")
-        ).map(AsRef::as_ref)
+        self.cargo_exe.get_or_try_init(|| {
+            fn from_current_exe() -> CargoResult<PathBuf> {
+                // Try fetching the path to `cargo` using env::current_exe().
+                // The method varies per operating system and might fail; in particular,
+                // it depends on /proc being mounted on Linux, and some environments
+                // (like containers or chroots) may not have that available.
+                env::current_exe()
+                    .and_then(|path| path.canonicalize())
+                    .map_err(CargoError::from)
+            }
+
+            fn from_argv() -> CargoResult<PathBuf> {
+                // Grab argv[0] and attempt to resolve it to an absolute path.
+                // Path::canonicalize succeeds only when the path is already absolute,
+                // or it's a relative path accessible from the current directory.
+                env::args_os()
+                    .next()
+                    .ok_or(CargoError::from("no argv[0]"))
+                    .map(PathBuf::from)
+                    .and_then(|argv0| argv0.canonicalize().or_else(|_| probe_path(argv0)))
+            }
+
+            fn probe_path(argv0: PathBuf) -> CargoResult<PathBuf> {
```

Comment:
Something similar happens in `execute_external_subcommand`, but I don't think we can reasonable extract common functionality here.